jep/237/README.adoc

```diff
@@ -65,55 +65,113 @@ endif::[]
+
+For these reasons it is not currently intended to add any automated code level checks into the ecosystem.
+
+=== Unit Tests
+
+As mentioned earlier it would be possible to unit test some behavioral differences.
+
+Using Unit tests to discover FIPS compliance issues would need several things:
+
+* The JVM would need to be appropriately configured for FIPS.
+* All test code and test dependencies would also need to be FIPS-140 compliant to prevent test infrastructure failures impacting the test.
+* Tests for code that is disabled when runung in FIP-140 would need to be disabled for these tests to not provide false positive issues
+
+The `maven-surefire` plugin used for Unit tests uses a flat classpath for unit tests, and is unaware of Jenkins' hierachical classloader and ability to have optional dependencies, or the ability to insert libraries into the classpath at runtime (which is used by the https://github.com/jenkinsci/bouncycastle-api-plugin/blob/2.30.1.77-225.v26ea_c9455fd9/src/main/java/jenkins/bouncycastle/api/BouncyCastlePlugin.java#L36-L67[`bouncycastle-api`] plugin).
+The assumption of the classpath would cause many regular unit tests to pass when they may be expected to fail.
+
+In addition to the above requirements and limitation, the unit tests would only be able to detect some concrete violations.
+They for example would not be able to fail if a plugin implemented its own encryption routines, or used custom implementations from another library (e.g. `bcrypt`)
+
+For these reasons it is not currently intended to add unit test level checks into the ecosystem.
+
+=== Acceptance Tests
+
+The https://github.com/jenkinsci/acceptance-test-harness[`acceptance test harness`] would appear to be a suitable place for tests to detect FIPS regressions and issues.
+
+As the ATH generally interacts with "external" systems (usually by way of a container) to check for regressions the environment should be a FIPS enabled environment, that includes the docker host as well as the system running Jenkins.
+This system can be independent of the host running the test (junit) code.
+The regression/test suite would need to be tagged with which plugins to test (or not test) and any agents launched for builds as part of the test would also need to be FIPS enabled.
+
+As with the other automated checks this can detect issues and regressions in features, but can not validate that the code is FIPS-140 compliant.
+These tests are generally the most expensive to run and coverage of plugins/features is variable.
+Whilst the value of finding regressions is certainly important, the ATH should be checking features, and if the feature works outside of FIPS (in a run of the current ATH) then the unit testing of the switched can provide this check, so this leaves the value at checking the JVM configuration is not causing regressions.
+As the JVM configuration is out of scope of the Jenkins project as outlined above, testing this configuration is of little value.
```

Comment:
Configuring a FIPS-compliant JVM for running ATH tests in CI does not seem insurmountable from an infrastructure perspective, assuming that all of the libraries needed to run such a JVM are free and open-source software, and doing so seems like a valuable method of ensuring that FIPS regressions are not inadvertently introduced in the future.

Reply:
configuring a JVM alone is not insurmountable,  but then I do not see what value we would get from this expensive testing for the cost of configuring and running it.

A FIPS compliant JVM does not necessarily prevent non FIPS compliant code from running, as such you could regress the FIPS compliant plugin code and the ATH may very well not detect it.

For example - anything launching a SSH Agent today[1] for tests in the ATH will work - but SSH Agents use `trilead-api` and this is so far from being FIPS compliant it is a non starter to make it compliant.

So what it can catch is if some code that was only exercised when Jenkins was started in FIPS mode regressed - but that would be better tested as a `RealJenkinsRule` or other unit test as [`HudsonPrivateSecurityRealmFIPSTest`](https://github.com/jenkinsci/jenkins/blob/master/test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java). 
But we can achieve most of the same just by starting Jenkins with the FIPS flag to catch these styles of regression should there be a valid reason to perform this style of testing over an ATH style test.

Iff we wanted to run ATHs for jenkins core and plugins then in addition to a fips compliant JVM we would also need to:
1) tag the ATHs we want to run by analysing them and adding them to a JUnit category or other.
2) ensure that the host that runs the docker deamon is also FIPS compliant (the kernel flag is inherited by containers and some OS/Libraries look at that flag to change behaviour at runtime) (requires dedicated k8s `node pool` from ci.jenkins.io) 
3) ensure any service specific configuration for the docker fixture is set correctly (e.g. sshd would need to be restricted to FIPS compliant algorithms in the SSHd server/OS combination that the service supports)
4) other minor tweaks/updates to the ATH framework (e.g.  [`DumbSlave`](https://github.com/jenkinsci/acceptance-test-harness/blob/5bee05f80e9118e2fb4f530d49e9d50a623422b4/src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java#L66) ) 
5) misc updates to any tests as required, e.g.  launching java (e.g. https://github.com/jenkinsci/acceptance-test-harness/blob/5bee05f80e9118e2fb4f530d49e9d50a623422b4/src/test/java/plugins/PipelineTest.java#L74)



I attempted to summarise this on [line 166](https://github.com/jenkinsci/jep/pull/407/files#diff-5c203956c4792d2dd52ffdb4e424e6c3bf99031d8af0ecabc399d7deed3ad1ddR166) and the preceding lines,

> Whilst the value of finding regressions is certainly important, the ATH should be checking features, and if the feature works outside of FIPS (in a run of the current ATH) then the unit testing of the switched can provide this check, so this leaves the value at checking the JVM configuration is not causing regressions.

but it appears I may have not done so effectively?

There is an additional concern that I did not outline, in that it can bring a false sense of security that it could bring ("the ATH passed this must be FIPS compliant").



[1] trilead-api probably needs to make this plugin just throw a [`BootFailure`](https://javadoc.jenkins.io/hudson/util/BootFailure.html) when run with the FIPS flag...